src/app/archive/archive_lib/processor.ml

```diff
@@ -638,24 +762,57 @@ module Coinbase = struct
         in
         Conn.find
           (Caqti_request.find typ Caqti_type.int
-             "INSERT INTO internal_commands (type, receiver_id, fee, token, \
-              hash) VALUES (?, ?, ?, ?, ?) RETURNING id")
+             {sql| INSERT INTO internal_commands
+                    (type, receiver_id, fee, token, hash)
+                   VALUES (?, ?, ?, ?, ?)
+                   RETURNING id
+             |sql})
           { receiver_id
           ; amount= Coinbase.amount t |> Currency.Amount.to_int
           ; hash= transaction_hash |> Transaction_hash.to_base58_check }
 end
 
-module Block_and_Internal_command = struct
+module Block_and_internal_command = struct
   let add (module Conn : CONNECTION) ~block_id ~internal_command_id
       ~sequence_no ~secondary_sequence_no =
     Conn.exec
       (Caqti_request.exec
          Caqti_type.(tup4 int int int int)
-         "INSERT INTO blocks_internal_commands (block_id, \
-          internal_command_id, sequence_no, secondary_sequence_no) VALUES (?, \
-          ?, ?, ?)")
+         {sql| INSERT INTO blocks_internal_commands
+                (block_id, internal_command_id, sequence_no, secondary_sequence_no)
+                VALUES (?, ?, ?, ?)
+         |sql})
       (block_id, internal_command_id, sequence_no, secondary_sequence_no)
 
+  let find (module Conn : CONNECTION) ~block_id ~internal_command_id
+      ~sequence_no ~secondary_sequence_no =
+    Conn.find_opt
+      (Caqti_request.find_opt
+         Caqti_type.(tup4 int int int int)
+         Caqti_type.string
+         {sql| SELECT 'exists' FROM blocks_internal_commands
+               WHERE block_id = $1
+               AND internal_command_id = $2
+               AND sequence_no = $3
+               AND secondary_sequence_no = $4
```

Comment:
Cool, TIL! (Haven't tested this, but I assume you have)

Reply:
Yes, the `exists` trick is confirmed to work with Postgresql.